### ISSUE TITLE
weekly_schedule error?

### ISSUE THREAD

**Issue by Devineavenger** [shane-mason/FieldStation42#11]
Hello, thank you for the work. I've wanted something like this, but never had the time to look into it.

I'm trying to build a small test channel. It only has 40-20 min episode in one "test" folder and I've added it to the:
"clip_shows" : ["test"]
but I keep getting this error message when trying to build channel:

DEBUG:kids:Making schedule for monday
DEBUG:kids:Making Slot: 6
DEBUG:kids:Making Slot: 7
DEBUG:kids:Slot tag: test
Traceback (most recent call last):
  File "/home/user/tv-station/station_42.py", line 259, in <module>
    schedule = station.make_weekly_schedule()
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/user/tv-station/station_42.py", line 202, in make_weekly_schedule
    schedule[day_name] = self.make_daily_schedule(day_name)
                         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/user/tv-station/station_42.py", line 228, in make_daily_schedule
    schedule[slot] = self.make_clip_hour(tag)
                     ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/user/tv-station/station_42.py", line 143, in make_clip_hour
    candidate=self.catalog.find_candidate(tag, remaining_time)
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/user/tv-station/fs42/catalog.py", line 142, in find_candidate
    raise(Exception(err))
Exception: Could not find candidate video for tag=test under 860.7400000000005 in len - maybe add some shorter content?

I've read the wiki.
What am I missing?

Thank you!

**Comment by shane-mason** [comment]
I can help here - happy to. It found enough shows to get it up to over 45 minutes, but couldn't find a filler for that last 14 minutes. If the content is 20 to 40 minutes, it doesn't have to be a clip show - 20 minutes is interpreted as the main content for a 30 minute block (so it will put two of them together to make an hour) and 40 minutes as the main content for a 60 minute block. Try it as a regular show, or follow the suggestion in the error message to `add some shorter content`

You will also need some content in your commercial and bump folders - it will try to hit the 60 minute mark pretty closely, so you want a combination of lengths for 15 to 90 seconds in your commercials (can be all 15-30 seconds) and then some in your bumps running from 5 to 90 seconds - this will let if find content to get the padding right.

Let me know if this works.

**Comment by shane-mason** [comment]
Not easily - it will try to fill the space with commercials and bumps to even out at half hour and hour marks - just like a real broadcast TV station would.  One way to do it would be to have all content that is exactly an hour or half hour, then there is no space to be filled. This seems like a simple feature, on the surface, but would become more complex because I built the scheduler around the concept of 'day parting' - and it looks at the world through one-hour increments. In your scenario, when a show ended and a new show begins would be completely dependent on the length of each content video. Can you tell me more about the use case? Marathons have commercials - even HBO generally has things start on the hour or half hour with bump content in between.

**Comment by Devineavenger** [comment]
My kids are small, like really small, and there are no commercials I would like to expose them to. They have a 60 to 90 min daily allowance for tv. I don't mind the rest of the day, as I will be the only one who will look/listen to it while I work. But for those let's say 2h, it would be nice to run cartoons without commercials. I've actually got a lot of commercials that I want to integrate on the channel, but more than half of them are 80's/90's 14+ commercials, not for 5yo... And even if I would split the content into 2 channels, 1 for me and a 2nd for them, i still wouldn't want commercials for that channel.

**Comment by shane-mason** [comment]
I would have to think about this for a bit - the best way to implement. Right now, you could do the following: replace the commercials with public service kinds of things. Like 'The more you know..." kind of clips that kids like. That keeps the 'commercials' out. I can pretty easily make it so you can specify no commercials or bumps, but what do we do with the remaining time in the slot? At some point, its just a random video playlist - which feels outside the scope of this particular project. We can do marathons right now, just by only having one show in your folder. We do need to be able to support 'no commercials' in a block, but we need time fillers. Think of how PBS does it: they have shows that range from 50 to 59 minutes or so and then fill the last part with previews and donation requests. HBO and movie channels are the same way.

**Comment by Devineavenger** [comment]
Could work. Last question. Can you make it so that you can specify 2-3 time slots that act as 1? Then you wouldn't need to fill up everything only the last remains of that 3h timeslot... that way, you can choose on which "hour" you like the filler (it's sort of a compromise to the whole "no commercials" thing)

**Comment by Devineavenger** [comment]
Started looking through the code. I have a solution (seeing that "commercial break" is out of whack anyway) that I don't have time, nor knowledge of how to implement.

Instead of

            if not len(matches):
                err = f"Could not find candidate video for tag={tag} under {seconds} in len - maybe add some shorter content?"
                raise(Exception(err))
            print(self._lowest_count(matches))
            return self._lowest_count(matches)


there should be something like 
choose random commercial, select x amount of remaining seconds and append to fill the last few seconds

my_clip = VideoFileClip(choose.random((com_tag))
my_clip.set_duration(seconds).append()

Maybe something like that? Again, sorry, I have no idea how python works... But hopefully you get the idea. That way it's always going to make a full block.

**Comment by shane-mason** [comment]
Want to note that I appreciate you keep working at this and providing me with feedback. Its helpful to see where its breaking for people.

**Comment by Devineavenger** [comment]
Always glad to help. Sorry I couldn't implement it... But learning python is not on my list right now. I can also guide you for the commercial break if you want. There is a lot of good code already in there that you can use, you just need to implement a few new things. And move some code around... Anyway, we can talk for that on the other issue. Happy coding :D

**Comment by shane-mason** [comment]
For sure - I'm pretty familiar with the code base ;)

**Comment by shane-mason** [comment]
Okay - I just committed an update that I think should help resolve this. I took a slightly different approach - I introduced two new types of exceptions and then refactored all the fill code to happen in a single function:

```
    def flexi_break(self, fill_time):
        reels = []
        remaining_time = fill_time

        keep_going = True

        while keep_going:
            try:
                fill = self.catalog.find_filler(remaining_time)
                remaining_time-=fill.duration
                reels.append(fill)
            except NoFillerContentFound as e:
                self._l.error("No commercials or bumps found - please add content to commercial and bump folders or check your configuration")
                raise e
            except MatchingContentNotFound:
                if(remaining_time > 30):
                    self._l.warning(f"Couldn't find commercials or bumps less than {remaining_time} seconds - FieldStation42 requires fill content to simulate accurate schedules.")
                keep_going = False

        return reels
```

The rationale behind this approach is that if you have a good bump catalog with items that are only a few seconds long then it can get really close to perfect timing.  Its also flexible, in that it will still let you have a gaps and just warn you. The downside is, it will leave gaps if you don't have good fill content.

Since the purpose of this project is to simulate broadcast TV, you really do need some 30 second fill content for it to work - and some 1, 2, 3, 5 and 10 second content if you want it to feel accurate.

**Comment by Devineavenger** [comment]
That might just work and it's actually a really good step towards "fixing" the commercial break with proper lead in and lead out bumpers :D I'll test when I get some time later and let you know.

**Comment by Devineavenger** [comment]
just added new branch and made the scripts, will also move the .py files

**Comment by shane-mason** [comment]
Thanks - I'll check out the commits tomorrow - thanks for the input. Sorry that the structure bugs you, but I am generally okay with it since they are the 'executable' pieces and not the library.

**Comment by Devineavenger** [comment]
Done. I know what "users" do :D That's why it bugged me... This is simpler. 4 scripts. clearly labeled. You also need to update the docs in the respective places. Thank you

**Comment by shane-mason** [comment]
Dude - you're giving me instructions like I work for you!  ;)

While a  program is in heavy development, it makes sense to keep it simple for the person(s) developing and then later (when the project is mature), we can build installable packages that will make it simpler for users. The file layout as specifically designed to make it easier for me to develop - I refactored that just a few weeks ago and I am very comfortable with how the files are currently structured. Later, there can be an installable package that works better for users.

Noting that I've resolved your issue with commercial breaks, I am resolving this ticket. Feel free to open new ones if you see new issues or have a feature request.
